ReactSkia/sdk/NopollWebsocket.cpp

```diff
@@ -0,0 +1,292 @@
+/*
+* Copyright (C) 1994-2022 OpenTV, Inc. and Nagravision S.A.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+#include <nopoll.h>
+
+#include "ReactSkia/utils/RnsLog.h"
+#include "NopollWebsocket.h"
+
+using namespace std;
+namespace facebook {
+namespace react {
+
+NopollWebsocket* NopollWebsocket::sharedNopollWebsocket_{nullptr};
+std::mutex NopollWebsocket::requestLock_;
+
+NopollWebsocket::NopollWebsocket() {
+  ctx_ = nopoll_ctx_new ();
+  sem_init(&startDownloadingThread_, 0, 0);
+  wsMessageThread_ = std::thread([this]() {
+    std::thread t1(&NopollWebsocket::pollThread,this);
+    int error_code;
+    while (ctx_) {
+      // wait for ever
+      error_code = nopoll_loop_wait (ctx_, 0);
+      if (error_code == -4)
+        RNS_LOG_ERROR(" io waiting mechanism, errno=" << errno);
+      // recover by just calling io wait engine
+      // try to limit recoveries to avoid infinite loop
+    }
+  });
+}
+
+NopollWebsocket* NopollWebsocket::sharedNopollWebsocket() {
+  std::lock_guard<std::mutex> lock(requestLock_);
+  if(sharedNopollWebsocket_ == nullptr)
+    sharedNopollWebsocket_ = new NopollWebsocket();
+  return sharedNopollWebsocket_;
+}
+
+NopollWebsocket::~NopollWebsocket() {
+  nopoll_ctx_unref(ctx_);
+  std::lock_guard<std::mutex> lock(requestLock_);
+  if(this == sharedNopollWebsocket_)
+    sharedNopollWebsocket_ = nullptr;
+}
+
+void NopollWebsocket::multiConn(){
+  NopollRequest* nopollRequest = nullptr;
+  while(true) { // forever
+    requestQueueLock_.lock();
+    if(RequestQueue_.size() !=0) {
+      nopollRequest =RequestQueue_.front();
+      if(!nopollRequest)
+        continue;
```

Comment:
dont we need to unlock before calling continue, else we wont be able to acquire the lock in the next loop

Reply:
updated